### ISSUE TITLE
Suggestion: Allow non-field targets

### ISSUE THREAD

**Issue by MingweiSamuel** [chancancode/rust-delegate#14]
Example:
```rs
use std::sync::Mutex;
use delegate::delegate;

#[derive(Clone, Debug)]
struct AtomicStack<T> {
    inner: Mutex<Vec<T>>,
}

impl<T> AtomicStack<T> {
    pub fn new() -> AtomicStack<T> {
        Stack { inner: Mutex::new(vec![]) }
    }

    delegate! {
        target self.inner.lock().unwrap() {
            /// The number of items in the stack
            #[target_method(len)]
            pub fn size(&self) -> usize;

            /// Whether the stack is empty
            pub fn is_empty(&self) -> bool;

            /// Push an item to the top of the stack
            pub fn push(&self, value: T);

            /// Remove an item from the top of the stack
            pub fn pop(&self) -> Option<T>;

            /// Accessing the top item without removing it from the stack
            #[target_method(last)]
            pub fn peek(&self) -> Option<&T>;

            /// Remove all items from the stack
            pub fn clear(&self);
        }
    }
}
```

**Comment by Kobzol** [comment]
Hi! :) This is actually pretty easy to implement and it would enable delegating wrappers, which is nice. Although writing arbitrary expressions/method calls (we can also allow calling global functions) inside a macro feels a bit unintuitive to me.

Here's what I don't like:
- When you make an error in the call expression, you get an error message pointing to the expression. However if the error occurs in the generated delegation method, it will still be matched back to the call expression, so it is hard to find out what is wrong. Try to use the Mutex example with a delegated method that takes `self` by value to see what I mean. I'll check if this can be fixed.
- The call expression looks like it is invoked only once, but it is actually copy pasted inside each delegated method and it will be called everytime the delegated method is invoked. Therefore the expression should ideally not have any side effects. Though in Rust it's relatively safe because side effects are kind of difficult to create with such a function call, so maybe it's not a big deal.

I suppose that if it will be optional, there's not much harm to be done. I implemented a prototype of it here: https://github.com/chancancode/rust-delegate/tree/expr. Please test it out if it works for your use cases.

**Comment by MingweiSamuel** [comment]
I confirmed that this works for my (relatively simple) use case. I definitely understand the concerns though

**Comment by Kobzol** [comment]
Closed by #15.
